src/pydicom/filewriter.py

```diff
@@ -303,14 +303,48 @@ def write_numbers(fp: DicomIO, elem: DataElement, struct_format: str) -> None:
         raise OSError(f"{str(e)}\nfor data_element:\n{str(elem)}")
 
 
+def _check_encaps_pixel_data(fp: DicomIO, value: bytes):
+    # valid pixel data with undefined length shall contain encapsulated
+    # data, e.g. sequence items - raise ValueError otherwise (see #238)
+
+    encap_item = b"\xfe\xff\x00\xe0"
+    if not fp.is_little_endian:
+        # Non-conformant endianness
+        encap_item = b"\xff\xfe\xe0\x00"
+    if not value.startswith(encap_item):
+        raise ValueError(
+            "(7FE0,0010) Pixel Data has an undefined length indicating "
+            "that it's compressed, but the data isn't encapsulated as "
+            "required. See pydicom.encaps.encapsulate() for more "
+            "information"
+        )
+
+
 def write_OBvalue(fp: DicomIO, elem: DataElement) -> None:
     """Write a data_element with VR of 'other byte' (OB)."""
-    if len(elem.value) % 2:
-        # Pad odd length values
-        fp.write(cast(bytes, elem.value))
-        fp.write(b"\x00")
+    value_length = 0
+    encaps_bytes: bytes = b""
+    if elem.is_buffered:
+        # stream data instead of reading the whole value at once
+        for chunk in elem.value_generator():
+            # need to store the first few bytes to check if the pixel data
+            # is encapsulated or not
+            if not encaps_bytes:
+                encaps_bytes = chunk[0:4]
+
+            fp.write(chunk)
+
+        value_length = elem.bytes_read_from_buffer
     else:
         fp.write(cast(bytes, elem.value))
+        value_length = len(elem.value)
+
+    if value_length % 2:
+        # Pad odd length values
+        fp.write(b"\x00")
+
+    if elem.is_undefined_length and elem.tag == 0x7FE00010:
+        _check_encaps_pixel_data(fp, cast(bytes, elem.value))
```

Comment:
I'm not sure if I like this being under `write_OB`, seems like something that only needs to be checked once per dataset, not on every OB element.

Shouldn't it also run when writing OW?

Reply:
Also, am I wrong in thinking that you're only writing OB buffer? Wouldn't you also want to write an OW buffer if you want to work with *Pixel Data*?

Reply:
Hmm, have I misunderstood or is the OW element buffer writing missing?

Reply:
It is missing for right now - I was just focusing on getting one VR working to begin with. I did intend to use _check_encaps_pixel_data in each of the byte writing functions though.

I can move it back to where it was but it was easiest to move it to where the reading actually happens so that I could store the first 4 bytes without having to seek backwards on the elem's buffer.

Reply:
No problem